test/e2e_node/image_pull_test.go

```diff
@@ -0,0 +1,286 @@
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	"k8s.io/kubernetes/test/e2e/nodefeature"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	admissionapi "k8s.io/pod-security-admission/api"
+	"k8s.io/utils/ptr"
+)
+
+var _ = SIGDescribe("Pull Image", framework.WithSerial(), nodefeature.MaxParallelImagePull, func() {
+
+	f := framework.NewDefaultFramework("parallel-pull-image-test")
+	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
+
+	httpdImage := imageutils.GetE2EImage(imageutils.Httpd)
+	httpdNewImage := imageutils.GetE2EImage(imageutils.HttpdNew)
+	var testpod, testpod2 *v1.Pod
+
+	ginkgo.Context("parallel image pull with MaxParallelImagePulls=5", func() {
+
+		tempSetCurrentKubeletConfig(f, func(ctx context.Context, initialConfig *kubeletconfig.KubeletConfiguration) {
+			initialConfig.SerializeImagePulls = false
+			initialConfig.MaxParallelImagePulls = ptr.To[int32](5)
+		})
+
+		ginkgo.BeforeEach(func(ctx context.Context) {
+			testpod = &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "httpd",
+					Namespace: f.Namespace.Name,
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{{
+						Name:            "httpd",
+						Image:           httpdImage,
+						ImagePullPolicy: v1.PullAlways,
+					}},
+					RestartPolicy: v1.RestartPolicyNever,
+				},
+			}
+			testpod2 = &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "httpd2",
+					Namespace: f.Namespace.Name,
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{{
+						Name:            "httpd-new",
+						Image:           httpdNewImage,
+						ImagePullPolicy: v1.PullAlways,
+					}},
+					RestartPolicy: v1.RestartPolicyNever,
+				},
+			}
+
+			ginkgo.By("cleanup images")
+			_ = RemoveImage(testpod.Spec.Containers[0].Image)
+			_ = RemoveImage(testpod2.Spec.Containers[0].Image)
+		})
+		ginkgo.AfterEach(func(ctx context.Context) {
+			ginkgo.By("cleanup pods")
+			if testpod != nil {
+				deletePodSyncByName(ctx, f, testpod.Name)
+			}
+			if testpod2 != nil {
+				deletePodSyncByName(ctx, f, testpod2.Name)
+			}
+		})
+
+		ginkgo.It("should pull immediately if no more than 5 pods", func(ctx context.Context) {
+			node := getNodeName(ctx, f)
+			testpod.Spec.NodeName = node
+			testpod2.Spec.NodeName = node
+
+			pod := e2epod.NewPodClient(f).Create(ctx, testpod)
+			pod2 := e2epod.NewPodClient(f).Create(ctx, testpod2)
+			err := e2epod.WaitForPodCondition(ctx, f.ClientSet, f.Namespace.Name, pod.Name, "Failed", 30*time.Second, func(pod *v1.Pod) (bool, error) {
+				if pod.Status.Phase == v1.PodRunning || pod.Status.Phase == v1.PodFailed {
+					return true, nil
+				}
+				return false, nil
+			})
+			framework.ExpectNoError(err)
+			err = e2epod.WaitForPodCondition(ctx, f.ClientSet, f.Namespace.Name, pod2.Name, "Failed", 30*time.Second, func(pod *v1.Pod) (bool, error) {
+				if pod.Status.Phase == v1.PodRunning || pod.Status.Phase == v1.PodFailed {
```

Comment:
Why v1.PodFailed status is ok? Shouldn't we expect running pod?

Reply:
Change to running.